### DISCUSSION TITLE
error-stack: Why does `error_stack::Report::frames()` produce a breadth-first iterator?

### DISCUSSION THREAD

**Discussion by mkatychev** [hashintel/hash#2758] [error-stack (Help)] [answered]
Given the rust code below:
```rust
use error_stack::Report;

#[derive(Debug, thiserror::Error)]
#[error("A")]
pub struct AError;

#[derive(Debug, thiserror::Error)]
#[error("C")]
pub struct CAttachment;

#[derive(Debug, thiserror::Error)]
#[error("D")]
pub struct DAttachment;

fn main() -> Result<(), Report<AError>> {
    let report = Report::new(CAttachment)
        .attach_printable(Report::new(DAttachment).attach_printable("E"))
        .change_context(AError)
        .attach_printable("B");

    for frame in report.frames() {
        println!("{frame:?}");
    }
    println!();
    Err(report)
}
```

The output below is given:
```
Frame { attachment: "B" }
Frame { .. }
Frame { context: AError }
Frame { attachment: D
├╴at src/lib.rs:17:27
╰╴E }
Frame { .. }
Frame { context: CAttachment }

Error: A
├╴at src/lib.rs:18:10
├╴B
│
╰─▶ C
    ├╴at src/lib.rs:16:18
    ╰╴D
```

Why does the `Display` on `Report<AError>` produce the right Depth-first ordering inline with the [`error_stack::iter::Frames`](https://docs.rs/error-stack/latest/error_stack/iter/struct.Frames.html#example) documentation:
```
A
├╴...
├╴B
│
╰─▶ C
    ├╴...
    ├╴D
    ╰ E (hidden)
```

While calling `Report::frames()::iter` produces a breadth first iteration:
```
B
A
D
E
C
```

**Comment by TimDiekmann** [comment]
This looks like a bug to me, cc @indietyp. Thank you for reporting this!

**Reply by indietyp** [reply]
That's interesting, especially considering that we're testing exactly that case: https://github.com/hashintel/hash/blob/main/libs/error-stack/tests/test_iter.rs#L23-L65

Will try to figure out what is causing the divergence.

**Comment by indietyp** [comment]
Okay, I think I have an answer, this is quite peculiar, but basically what's happening is that you're constructing:

```rust
    let report = Report::new(CAttachment)
        .attach_printable(Report::new(DAttachment).attach_printable("E"))
        .change_context(AError)
        .attach_printable("B");
```

leads to the following tree:

```
B
|
A
|
D
|
C
```

So the output will be: `B A D C`, you are not using a context, so when attaching a printable you are basically attaching `D\nE`, which leads to:

```
B
A
D
E
C
```

**Reply by indietyp** [reply]
To attach a report "correctly", you would need to use `extend()` or `extend_one()`

**Reply by TimDiekmann** [reply]
Thanks, @indietyp, I also just figured out why the behavior is as it is 😅 🚀

**Reply by mkatychev** [reply]
@indietyp  do you mind showing the `let report` above using `extend()`? That will allow me to mark a comment as accepted answer.

**Reply by indietyp** [reply, marked as answer]
Sure! Sorry for the long wait. I kind of forgot to answer 😅 

```rust
use error_stack::Report;

#[derive(Debug, thiserror::Error)]
#[error("A")]
pub struct AError;

#[derive(Debug, thiserror::Error)]
#[error("C")]
pub struct CAttachment;

#[derive(Debug, thiserror::Error)]
#[error("D")]
pub struct DAttachment;

fn main() -> Result<(), Report<AError>> {
    let mut report = Report::new(CAttachment);
	report.extend_one(Report::new(DAttachment).attach_printable("E"));
	
    let report = report
        .change_context(AError)
        .attach_printable("B");

    for frame in report.frames() {
        println!("{frame:?}");
    }
    println!();
    Err(report)
}
```

This should fix the problem (notice the call to `extend_one`)

**Reply by mkatychev** [reply]
Appreciate the follow-up
